### ISSUE TITLE
Consider home-grown replacement for cron-utils

### ISSUE THREAD

**Issue by axelfontaine** [kagkarlsson/db-scheduler#255]
In order to reduce the number of dependencies (the ideal number would be 0), the small subset of cron-utils that is being used could easily be replaced by a minimal home-grown solution.

Spring did something similar which you can find here and may even be able to re-use as-is given the license: https://github.com/spring-projects/spring-framework/blob/76b1c0f1fcf9b4720cbd2d3602289af24de3bf69/spring-context/src/main/java/org/springframework/scheduling/support/CronTrigger.java

**Comment by axelfontaine** [comment]
Note that this could easily be extended to the remaining two dependencies:

- your own micro-jdbc library could either be folded in or shaded
- slf4j could be meade optional behind your own logging facade that would fall back to java.util.logging if slf4j isn't present (that's how I did it with Flyway)

**Comment by kagkarlsson** [comment]
Hi! Thanks for the suggestion. I used a shaded version of cron-utils to avoid having to write the parser myself, saving the time and bugs. I agree that it is not ideal, since it contains more code than necessary. For example, it recently needed patching due to a CVE in cron-utils. On the other hand, if I replicate or copy the code manually, the maintenance cost will be a bit higher.. 

Sure, micro-jdbc can be shaded. The only reason I have not is that I figure no one else is using it :). But it might be a good idea to reduce dependency-graph noise 🤔 

As for `slf4j`, good point. If we disregard the possibility for auto-detection, would the best default for an embeddable library be java.util.logging perhaps, not slf4j? Maybe with a config-option to log directly via slf4j.. (and alternatives)

**Comment by axelfontaine** [comment]
> On the other hand, if I replicate or copy the code manually, the maintenance cost will be a bit higher..

Trade-offs indeed. An empty dependency graph is quite a major benefit though.

> But it might be a good idea to reduce dependency-graph noise

The simpler the better!

> Maybe with a config-option to log directly via slf4j

I would highly recommend making things "just work" out of the box when possible instead of adding configuration options as the added complexity they bring directly translates into a higher overhead for new users. In this case auto-detection is super simple. See https://github.com/flyway/flyway/blob/master/flyway-core/src/main/java/org/flywaydb/core/internal/util/FeatureDetector.java#L57

**Comment by kagkarlsson** [comment]
> I would highly recommend making things "just work" out of the box when possible instead of adding configuration options as the added complexity they bring directly translates into a higher overhead for new users. In this case auto-detection is super simple. See https://github.com/flyway/flyway/blob/master/flyway-core/src/main/java/org/flywaydb/core/internal/util/FeatureDetector.java#L57

Yeah, I checked it out, nice job. Did you run into any weird edge-cases or happy users all around?

**Comment by axelfontaine** [comment]
No weird stuff, you just need to be careful which classloader you use to check things, although that's mainly just an OSGi issue, and OSGi isn't widely used anyway. Other than that users really don't think about it as the library "just works" regardless of their environment.
